Math/NumberTheory/SmoothNumbers.hs

```diff
@@ -76,29 +80,43 @@ fromSmoothUpperBound n = if (n < 2)
                          then Nothing
                          else Just $ SmoothBasis $ takeWhile (<= n) $ map unPrime primes
 
--- | Generate an infinite ascending list of
--- <https://en.wikipedia.org/wiki/Smooth_number smooth numbers>
--- over a given smooth basis.
---
--- >>> import Data.Maybe
--- >>> take 10 (smoothOver (fromJust (fromList [2, 5])))
--- [1, 2, 4, 5, 8, 10, 16, 20, 25, 32]
-smoothOver :: Integral a => SmoothBasis a -> [a]
-smoothOver pl = foldr (\p l -> mergeListLists $ iterate (map (p*)) l) [1] (unSmoothBasis pl)
+-- | Helper used by @smoothOver@ (@Integral@ constraint) and @smoothOver'@
+-- (@Euclidean@ constraint) Since the typeclass constraint is just
+-- @Num@, it receives a @norm@ comparison function for the generated smooth
+-- numbers.
+-- This function relies on the fact that for any element of a smooth basis @p@
+-- and any @a@ it is true that @norm (a * p) > norm a@.
+-- This condition is not checked.
+smoothOver' :: forall a b . (Eq a, Num a, Ord b) => (a -> b) -> SmoothBasis a -> [a]
+smoothOver' norm pl =
+    foldr
+    (\p l -> mergeListLists $ iterate (map (p*)) l)
+    [1]
+    (nub $ unSmoothBasis pl)
   where
     {-# INLINE mergeListLists #-}
     mergeListLists      = foldr go1 []
       where
-        go1 :: Ord a => [a] -> [a] -> [a]
+        go1 :: [a] -> [a] -> [a]
         go1 (h:t) b = h:(go2 t b)
         go1 _     b = b
 
-        go2 :: Ord a => [a] -> [a] -> [a]
+        go2 :: [a] -> [a] -> [a]
         go2 a@(ah:at) b@(bh:bt)
-          | bh < ah   = bh : (go2 a bt)
+          | norm bh < norm ah   = bh : (go2 a bt)
           | otherwise = ah : (go2 at b) -- no possibility of duplicates
```

Comment:
Now it is worth to check for duplicates. Otherwise some smooth numbers may repeat. For example,

```haskell
> take 10 $ smoothOver $ Data.Maybe.fromJust $ fromList [2,4]
[1,2,4,4,8,8,16,16,16,32]
```

(It is also worth to add a test IMHO)

Reply:
@Bodigrim I really don't like the thought of using `nub ∈ O(N²)` here... I think it's worth a refactor to use a `Set` inside this loop.

Reply:
Nevermind, I spoke too soon. It was a quick fix in the line below this one, just another guard in the pattern match.